xwiki-platform-core/xwiki-platform-administration/xwiki-platform-administration-ui/src/main/resources/XWiki/ForgotUsername.xml

```diff
@@ -49,6 +49,10 @@
 
 #else
   #set($results = $services.query.hql(", BaseObject obj, StringProperty prop where obj.name = doc.fullName and obj.className = 'XWiki.XWikiUsers' and prop.id.id = obj.id and prop.id.name = 'email' and LOWER(prop.value) = ?").bindValues([$email.toLowerCase()]).execute())
+  ## If local user does not exist check global user
+  #if($results.size() == 0)
```

Comment:
The case where the current wiki is already the main wiki is missing. We should avoid running the query twice, for the main wiki.